src/mlpack/methods/ann/dag_network_impl.hpp

```diff
@@ -0,0 +1,865 @@
+  }
+
+  //setup layerInputs
+  for (size_t i = 1; i < sortedNetwork.size(); i++)
+  {
+    Layer<MatType>* currentLayer = sortedNetwork[i];
+    std::vector<Layer<MatType>*> parents = adjacencyList[currentLayer];
+    size_t numParents = parents.size();
+    if (numParents <= 0)
+    {
+      // TODO: should be caught in CheckGraph
+      throw std::logic_error("DAGNetwork::InitializeForwardPassMemory(): Non-input nodes must have atleast one parent node.");
+    }
+
+    if (numParents == 1)
+    {
+      size_t inputIndex = sortedIndices[parents.front()];
+      MakeAlias(layerInputs[i-1], layerOutputs[inputIndex], layerOutputs[inputIndex].n_rows, layerOutputs[inputIndex].n_cols, 0);
+    }
+    else
+    {
+      size_t concatSize = currentLayer->InputDimensions()[0];
+      for (size_t j = 1; j < currentLayer->InputDimensions().size(); j++)
+        concatSize *= currentLayer->InputDimensions()[j];
+      MakeAlias(layerInputs[i-1], layerOutputMatrix, concatSize, batchSize, offset);
+      offset += concatSize;
+    }
+  }
+}
+
+template<typename OutputLayerType,
+         typename InitializationRuleType,
+         typename MatType>
+void DAGNetwork<
+    OutputLayerType,
+    InitializationRuleType,
+    MatType
+>::Forward(const MatType& input, MatType& results)
+{
+  CheckNetwork("DAGNetwork::Forward()", input.n_rows);
+
+  networkOutput.set_size(sortedNetwork.back()->OutputSize(), input.n_cols);
+
+  // equivalent to MultiLayer->Forward(), but with a concat.
+  if (sortedNetwork.size() > 1)
+  {
+    InitializeForwardPassMemory(input.n_cols);
+    sortedNetwork.front()->Forward(input, layerOutputs.front());
+    for (size_t i = 1; i < sortedNetwork.size(); i++)
+    {
+      std::vector<Layer<MatType>*> parents = adjacencyList[sortedNetwork[i]];
+
+      if (parents.size() > 1)
+      {
+        std::vector<CubeType> parentOutputAliases(parents.size());
+        size_t axis = layerAxes[sortedNetwork[i]];
+
+        size_t rows = 1;
+        for (size_t j = 0; j < axis; j++)
+          rows *= sortedNetwork[i]->InputDimensions()[j];
+
+        size_t slices = input.n_cols;
+        for (size_t j = axis + 1; j < sortedNetwork[i]->InputDimensions().size(); j++)
+          slices *= sortedNetwork[i]->InputDimensions()[j];
+
+
+        for (size_t j = 0; j < parents.size(); j++)
+        {
+          size_t cols = parents[j]->OutputDimensions()[axis];
+          MatType& parentOutput = layerOutputs[sortedIndices[parents[j]]];
+          MakeAlias(parentOutputAliases[j], parentOutput, rows, cols, slices);
+        }
+
+        CubeType inputAlias;
+        MakeAlias(inputAlias, layerInputs[i-1], rows, sortedNetwork[i]->InputDimensions()[axis], slices);
+
+        size_t startCol = 0;
+        for (size_t j = 0; j < parentOutputAliases.size(); j++)
+        {
+          const size_t cols = parentOutputAliases[j].n_cols;
+          inputAlias.cols(startCol, startCol + cols - 1) = parentOutputAliases[j];
+          startCol += cols;
+        }
```

Comment:
But if we are concatenating on the last axis (`axis == sortedNetwork[i]->InputDimensions().size() - 1`), then shouldn't we be able to do nothing at all?  In that case you could set up `layerOutputs` of the previous layer such that they are the same memory as `layerInputs[i]` (sorry if `i` is not the right index there but I think you get what I mean).

I think there could be some extra caveats on when you can apply this optimization, but in any case I think it is definitely worthwhile to implement as it can prevent an entire copy for every forward/backward/gradient pass.